### DISCUSSION TITLE
ACL: Can't get %c in topic to match in EMQX 5.0.10

### DISCUSSION THREAD

**Discussion by perbu** [emqx/emqx#9421] [Q&A] [answered]
Hi.

I'm trying to upgrade from 4.x to 5.0. Thing are looking good except this one ACL I have which I can't get to work.

In file it looks like this:
```
(a few other ACLs)
{allow, all, all, ["foo/v1/%c/#"]}.
{deny, all}
```
Loading this when a client tries to connect and publish I get the following:
```
2022-11-24T14:36:05.329512+00:00 [warning] clientid: 3136, ipaddr: {10,166,0,37}, line: 356, mfa: 
emqx_authz:authorize_non_superuser/5, msg: authorization_permission_denied, peername: 10.166.0.37:24163, 
source: file, topic: foo/v1/3136/metadata/connectivity/events/connectionup, username: undefined
```

Now I've tried replacing the %c with the clientid in question (3136) and this works fine. So I kind of suspect that something around how %c is handled has changed. I've tried putting the same rule in the built in database, but the results are the same. It works with a hard-coded client id, but fails when I replace it with %c.

I've tried to have a look at the authz stuff in the source, but Erlang looks pretty much greek to me. :-)

Anyone knows if works in 5.0.10?

**Comment by perbu** [comment]
Ok. So I tried the following on with a fresh Docker container.
I went in and edited the ACL file through the web interface so it looked like this:
```
{allow, {username, "^dashboard?"}, subscribe, ["$SYS/#"]}.
{allow, {ipaddr, "127.0.0.1"}, all, ["$SYS/#", "#"]}.
{deny, all, subscribe, ["$SYS/#", {eq, "#"}]}.
{allow, all, all, ["local/%c/#"]}.
{allow, {clientid, "admin"}, all, ["local/#"]}.
{deny, all}.
```
Now. I ran the following:
```
mosquitto_pub -i foo -t 'local/foo' -m 'test'
```
In the logs:
```
2022-11-24T19:05:57.505633+00:00 [warning] clientid: foo, ipaddr: {172,17,0,1}, line: 356, 
mfa: emqx_authz:authorize_non_superuser/5, msg: authorization_permission_denied, peername: 172.17.0.1:59336, 
source: file, topic: local/foo, username: undefined
2022-11-24T19:05:57.506063+00:00 [warning] clientid: foo, line: 663, mfa: emqx_channel:process_publish/2, 
msg: cannot_publish_to_topic, peername: 172.17.0.1:59336, reason: not_authorized, topic: local/foo
```

It seems %c handling is broken.

**Comment by perbu** [comment, marked as answer]
%c is replaced with  ${clientid}.

If I might make a couple of  suggestion:
1) I would make a note of this in the transition documentation. I don't think it is mentioned there.
2) I would issue a warning during ACL loading if %c or %u is used in the ACLs

**Reply by lafirest** [reply]
Thanks for your suggestions, the first point is obviously feasible
